FRAuth/FRAuth/AppIntegrity/FRAppAttestDomainModal.swift

```diff
@@ -46,52 +49,143 @@ struct FRAppAttestDomainModal: FRAppAttestation {
         }
+        
         guard let bundleIdentifier = bundleIdentifier, !bundleIdentifier.isEmpty else {
             throw FRDeviceCheckAPIFailure.invalidBundleIdentifier
         }
+        
+        guard let challengeUtf8 = challenge.data(using: .utf8), !challengeUtf8.isEmpty else {
+            throw FRDeviceCheckAPIFailure.invalidChallenge
+        }
+        
         let userClientData = [challengeKey: challenge,
                                bundleIdKey: bundleIdentifier]
         
         guard let jsonData = try? encoder.encode(userClientData) else {
             throw FRDeviceCheckAPIFailure.invalidClientData
         }
         
-        if !service.isSupported() {
-            throw FRDeviceCheckAPIFailure.featureUnsupported
-        }
+        return (challengeUtf8, jsonData)
         
+    }
+    
+    /// attestation
+    ///
+    /// - Parameter challenge: Challenge Received from server
+    /// - Parameter jsonData: jsonData Received from server
+    /// - Throws: `FRDeviceCheckAPIFailure and Error`
+    /// - Returns: FRAppIntegrityKeys
+    private func attestation(challenge: Data,
+                             jsonData: Data) async throws -> FRAppIntegrityKeys {
+        let keyId = try await service.generateKey()
+        let result = try await service.attest(keyIdentifier: keyId, clientDataHash: Data(SHA256.hash(data: challenge)))
+        let attestation = result.base64EncodedString()
+        return FRAppIntegrityKeys(attestKey: attestation,
+                                  assertKey: nil,
+                                  keyIdentifier: keyId,
+                                  clientDataHash: jsonData.base64EncodedString())
+    }
+    
+    /// assertion
+    ///
+    /// - Parameter challenge: Challenge Received from server
+    /// - Parameter jsonData: jsonData Received from server
+    /// - Parameter keyIdValue: keyIdValue from keychain
+    /// - Parameter attestationValue: attestationValue from keychain
+    /// - Throws: `FRDeviceCheckAPIFailure and Error`
+    /// - Returns: FRAppIntegrityKeys
+    private func assertion(challenge: Data,
+                           jsonData: Data,
+                           keyIdValue: String,
+                           attestationValue: String) async throws -> FRAppIntegrityKeys {
         do {
-            let keyIdentifier = try await service.generateKey()
-            let attest = try await service.attest(keyIdentifier: keyIdentifier, clientDataHash: Data(SHA256.hash(data: challengeUtf8)))
-            let assert = try await service.generateAssertion(keyIdentifier: keyIdentifier, clientDataHash: Data(SHA256.hash(data: jsonData)))
-            return FRAppIntegrityKeys(attestKey: attest.base64EncodedString(),
-                                      assertKey: assert.base64EncodedString(),
-                                      keyIdentifier: keyIdentifier,
+            let assertion = try await withRetry {
```

Comment:
Out of curiosity the `withRetry` how many times does it retry? And why are we using it here? Is there a chance of failure?

Reply:
it will retry twice, and if its failed , we have a catch block to handle a special scenario